File: backend/app/schemas/deal_history.py
from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, computed_field

from app.schemas.client import LeadSource

DealType = Literal["sold", "removed"]


class DealHistoryOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    listing_id: int | None
    deal_type: DealType

    brand: str
    model: str
    year: str
    vin: str | None

    seller_name: str
    seller_phone: str
    buyer_name: str | None
    buyer_phone: str | None
    seller_lead_source: LeadSource | None
    buyer_lead_source: LeadSource | None

    final_price: float | None
    purchase_price: float | None
    additional_expenses: float | None

    employee_name: str | None
    employee_id: int | None

    date_closed: date
    date_added: date | None
    created_at: datetime

    @computed_field
    @property
    def net_profit(self) -> float | None:
        if self.final_price is None or self.purchase_price is None or self.additional_expenses is None:
            return None
        return self.final_price - self.purchase_price - self.additional_expenses

    @computed_field
    @property
    def days_on_lot(self) -> int | None:
        if self.date_added is None:
            return None
        return (self.date_closed - self.date_added).days

File: frontend/app/types/dealHistory.ts
import type { ClientLeadSource } from '~/utils/clientOptions'

export type DealType = 'sold' | 'removed'

// A snapshot at the moment the deal closed — not a live view. Renaming a
// client or an employee later never updates these rows retroactively.
export interface DealHistoryEntry {
    id: number
    listing_id: number | null
    deal_type: DealType

    brand: string
    model: string
    year: string
    vin: string | null

    seller_name: string
    seller_phone: string
    buyer_name: string | null
    buyer_phone: string | null
    seller_lead_source: ClientLeadSource | null
    buyer_lead_source: ClientLeadSource | null

    final_price: number | null
    purchase_price: number | null
    additional_expenses: number | null
    net_profit: number | null

    employee_name: string | null
    employee_id: number | null

    date_closed: string
    date_added: string | null
    days_on_lot: number | null
    created_at: string
}
